calyx-opt/src/passes/compile_static.rs

```diff
@@ -75,99 +87,265 @@ fn make_guard_dyn(
+        builder.add_continuous_assignments(continuous_assigns.to_vec());
+        g.borrow_mut().assignments = group_assigns.to_vec();
+        g.borrow_mut().attributes =
+            early_reset_group.borrow().attributes.clone();
+        g
+    }
+}
+
+impl Visitor for CompileStatic {
+    fn start(
+        &mut self,
+        comp: &mut ir::Component,
+        sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        let sgroups: Vec<ir::RRC<ir::StaticGroup>> =
+            comp.get_static_groups_mut().drain().collect();
+        let mut builder = ir::Builder::new(comp, sigs);
+        // create "early reset" dynamic groups that never reach set their done hole
+        for sgroup in sgroups.iter() {
+            let mut sgroup_ref = sgroup.borrow_mut();
+            let sgroup_name = sgroup_ref.name();
+            let sgroup_latency = sgroup_ref.get_latency();
+            let sgroup_attributes = sgroup_ref.attributes.clone();
+            let sgroup_assigns = &mut sgroup_ref.assignments;
+            let g = self.make_early_reset_group(
+                sgroup_assigns,
+                sgroup_name,
+                sgroup_latency,
+                sgroup_attributes,
+                &mut builder,
+            );
+            // map the static group name -> early reset group name
+            // helpful for rewriting control
+            self.reset_early_map.insert(sgroup_name, g.borrow().name());
+            // group_rewrite_map helps write static_group[go] to early_reset_group[go]
+            // technically could do this w/ early_reset_map but is easier w/
+            // group_rewrite, which is explicitly of type `PortRewriterMap`
+            self.group_rewrite.insert(
+                ir::Canonical(sgroup_name, ir::Id::from("go")),
+                g.borrow().find("go").unwrap_or_else(|| {
+                    panic!("group {} has no go port", g.borrow().name())
+                }),
+            );
+        }
+
+        // rewrite static_group[go] to early_reset_group[go]
+        // don't have to worrry about writing static_group[done] b/c static
+        // groups don't have done holes.
+        comp.for_each_assignment(|assign| {
+            assign.for_each_port(|port| {
+                self.group_rewrite
+                    .get(&port.borrow().canonical())
+                    .map(Rc::clone)
+            })
+        });
+
+        comp.get_static_groups_mut().append(sgroups.into_iter());
+
+        Ok(Action::Continue)
+    }
+
+    /// Executed after visiting the children of a [ir::Static] node.
+    fn start_static_control(
+        &mut self,
+        sc: &mut ir::StaticControl,
+        comp: &mut ir::Component,
+        sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        // assume that there are only static enables left.
+        // if there are any other type of static control, then error out.
+        match sc {
```

Comment:
Might be the right place to use the new [`let`-`else` construct](https://rust-lang.github.io/rfcs/3137-let-else.html) in Rust:
```
let ir::StaticControl::Enable(s) = sc else {
  unreachable!(...)
};
// Remaning stuff that uses `s`
```